File: bindings/swift/Sources/Proven/SafeUrl.swift
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/// Safe URL parsing delegated to libproven FFI.
///
/// All parsing is performed by the formally verified Idris 2 core
/// via the Zig FFI bridge.

import CProven

/// Parsed URL components returned by SafeUrl.parse.
public struct ParsedUrl: Equatable, Sendable {
    public let scheme: String
    public let host: String
    public let port: UInt16?
    public let path: String
    public let query: String?
    public let fragment: String?
}

public enum SafeUrl {
    /// Parse a URL into its components.
    public static func parse(_ urlString: String) -> Result<ParsedUrl, ProvenError> {
        withStringBytes(urlString) { ptr, len in
            var result = proven_url_parse(ptr, len)
            if let error = ProvenError.fromStatus(result.status) {
                return .failure(error)
            }

            let scheme: String
            if let sPtr = result.components.scheme, result.components.scheme_len > 0 {
                scheme = String(cString: sPtr)
            } else {
                scheme = ""
            }

            let host: String
            if let hPtr = result.components.host, result.components.host_len > 0 {
                host = String(cString: hPtr)
            } else {
                host = ""
            }

            let port: UInt16? = result.components.has_port ? result.components.port : nil

            let path: String
            if let pPtr = result.components.path, result.components.path_len > 0 {
                path = String(cString: pPtr)
            } else {
                path = "/"
            }

            let query: String?
            if let qPtr = result.components.query, result.components.query_len > 0 {
                query = String(cString: qPtr)
            } else {
                query = nil
            }

            let fragment: String?
            if let fPtr = result.components.fragment, result.components.fragment_len > 0 {
                fragment = String(cString: fPtr)
            } else {
                fragment = nil
            }

            // Free the C-allocated URL components.
            proven_url_free(&result.components)

            return .success(ParsedUrl(
                scheme: scheme,
                host: host,
                port: port,
                path: path,
                query: query,
                fragment: fragment
            ))
        }
    }
}

File: bindings/purescript/Proven/SafeUrl.js
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// SafeUrl FFI implementation - delegates to libproven via ffi-napi.
// All computation performed in Idris 2 via the Zig FFI layer.

"use strict";

const { lib } = require("./FFI.js");

function readNullTermString(ptr, len) {
  if (ptr === null || ptr.isNull()) return "";
  return ptr.readCString(0);
}

exports.parseUrlImpl = function (s) {
  var buf = Buffer.from(s, "utf-8");
  var r = lib.proven_url_parse(buf, buf.length);
  if (r.status === 0) {
    var result = {
      status: 0,
      scheme: readNullTermString(r.components.scheme, r.components.scheme_len),
      host: readNullTermString(r.components.host, r.components.host_len),
      port: r.components.has_port ? r.components.port : -1,
      path: readNullTermString(r.components.path, r.components.path_len),
      query: readNullTermString(r.components.query, r.components.query_len),
      fragment: readNullTermString(
        r.components.fragment,
        r.components.fragment_len
      ),
    };
    lib.proven_url_free(r.components);
    return result;
  }
  return {
    status: r.status,
    scheme: "",
    host: "",
    port: -1,
    path: "",
    query: "",
    fragment: "",
  };
};
